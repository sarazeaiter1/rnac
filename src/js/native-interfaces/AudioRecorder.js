import { NativeAppEventEmitter, Platform } from 'react-native';

import AudioRecorderManager from './AudioRecorderManager';
import getNativePrepareRecordingAtPathArguments from './getNativePrepareRecordingAtPathArguments';

import nativeAppEventNames from '../constants/nativeAppEventNames';

const nativeEventListeners = {
  recordingProgressSubscription: null,
  recordingFinishedSubscription: null
};

const recordingState = {
  isRecording: false
};

const AudioRecorder = {
  onRecordingProgress: () => null,
  onRecordingFinished: () => null,
  init({ onRecordingProgress, onRecordingFinished }) {
    this.onRecordingProgress = onRecordingProgress;
    this.onRecordingFinished = onRecordingFinished;
  },
  destroy() {
    if (recordingState.isRecording) {
      recordingState.isRecording = false;
      this.stopRecording();
    }
  },
  bindNativeEventListeners() {
    this.unbindNativeEventListeners();
    nativeEventListeners.recordingProgressSubscription = NativeAppEventEmitter.addListener(
      nativeAppEventNames.recordingProgress,
      this.onRecordingProgress
    );
    nativeEventListeners.recordingFinishedSubscription = NativeAppEventEmitter.addListener(
      nativeAppEventNames.recordingFinished,
      this.onRecordingFinished
    );
  },
  unbindNativeEventListeners() {
    if (nativeEventListeners.recordingProgressSubscription !== null) {
      nativeEventListeners.recordingProgressSubscription.remove();
    }
    if (nativeEventListeners.recordingFinishedSubscription !== null) {
      nativeEventListeners.recordingProgressSubscription.remove();
    }
  },
  async startRecording({ recordingPath, recordingOptions }) {
    try {
      if (!recordingState.isRecording) {
        recordingState.isRecording = true;
        this.bindNativeEventListeners();
        const prepareRecordingAtPathArgs = getNativePrepareRecordingAtPathArguments(
          {
            recordingPath,
            recordingOptions: { recordingOptions }
          }
        );
        await AudioRecorderManager.prepareRecordingAtPath(
          prepareRecordingAtPathArgs.recordingPath,
          {
            ...prepareRecordingAtPathArgs.settings
          }
        );

        return AudioRecorderManager.startRecording();
      }
    } catch (err) {
      // eslint-disable-next-line
      console.warn(err);
    }
  },
  stopRecording: function() {
    if (recordingState.isRecording) {
      recordingState.isRecording = false;
      this.unbindNativeEventListeners();
      return AudioRecorderManager.stopRecording();
    }
  },
  pauseRecording: function() {
    return Platform.OS === 'ios'
      ? AudioRecorderManager.pauseRecording()
      : this.stopRecording();
  },
  checkAuthorizationStatus: AudioRecorderManager.checkAuthorizationStatus,
  requestAuthorization: AudioRecorderManager.requestAuthorization,
  removeListeners: function() {
    if (nativeEventListeners.recordingProgressSubscription)
      nativeEventListeners.recordingProgressSubscription.remove();
    if (nativeEventListeners.recordingFinishedSubscription)
      nativeEventListeners.recordingFinishedSubscription.remove();
  }
};

export default AudioRecorder;
