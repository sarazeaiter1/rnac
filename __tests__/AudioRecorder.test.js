// import React from 'react';
// import { Text, View, Button } from 'react-native';
// import renderer from 'react-test-renderer';
// import { shallow } from 'enzyme';
import AudioRecorder from '../src/js/native-interfaces/AudioRecorder';

describe('Recorder', () => {
  test('exports', () => {
    expect(AudioRecorder).toMatchSnapshot();
  });
  test('init', () => {
    AudioRecorder.init({
      onRecordingProgress: null,
      onRecordingFinished: null
    });
    expect(AudioRecorder).toMatchSnapshot();
  });

  test('destroy', () => {
    AudioRecorder.startRecording('test.aac', {
      recordingPath: 'et3/',
      recordingOptions: {
        SampleRate: 44100.0,
        Channels: 2,
        AudioQuality: 'High',
        AudioEncoding: 'ima4',
        OutputFormat: 'mpeg_4',
        MeteringEnabled: false,
        AudioEncodingBitRate: 32000
      }
    });
    AudioRecorder.destroy();
    expect(AudioRecorder).toMatchSnapshot();
  });
  test('pauseRecording', () => {
    try {
      AudioRecorder.pauseRecording();
    } catch (err) {
      expect(AudioRecorder).toMatchSnapshot();
    }
  });
  test('removeListeners recordingProgressSubscription', () => {
    AudioRecorder.bindNativeEventListeners();
    AudioRecorder.removeListeners();
    expect(AudioRecorder).toMatchSnapshot();
  });
});
