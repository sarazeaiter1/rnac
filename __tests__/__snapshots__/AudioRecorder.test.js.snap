// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recorder destroy 1`] = `
Object {
  "bindNativeEventListeners": [Function],
  "checkAuthorizationStatus": undefined,
  "destroy": [Function],
  "init": [Function],
  "onRecordingFinished": null,
  "onRecordingProgress": null,
  "pauseRecording": [Function],
  "removeListeners": [Function],
  "requestAuthorization": undefined,
  "startRecording": [Function],
  "stopRecording": [Function],
  "unbindNativeEventListeners": [Function],
}
`;

exports[`Recorder exports 1`] = `
Object {
  "bindNativeEventListeners": [Function],
  "checkAuthorizationStatus": undefined,
  "destroy": [Function],
  "init": [Function],
  "onRecordingFinished": [Function],
  "onRecordingProgress": [Function],
  "pauseRecording": [Function],
  "removeListeners": [Function],
  "requestAuthorization": undefined,
  "startRecording": [Function],
  "stopRecording": [Function],
  "unbindNativeEventListeners": [Function],
}
`;

exports[`Recorder init 1`] = `
Object {
  "bindNativeEventListeners": [Function],
  "checkAuthorizationStatus": undefined,
  "destroy": [Function],
  "init": [Function],
  "onRecordingFinished": null,
  "onRecordingProgress": null,
  "pauseRecording": [Function],
  "removeListeners": [Function],
  "requestAuthorization": undefined,
  "startRecording": [Function],
  "stopRecording": [Function],
  "unbindNativeEventListeners": [Function],
}
`;

exports[`Recorder pauseRecording 1`] = `
Object {
  "bindNativeEventListeners": [Function],
  "checkAuthorizationStatus": undefined,
  "destroy": [Function],
  "init": [Function],
  "onRecordingFinished": null,
  "onRecordingProgress": null,
  "pauseRecording": [Function],
  "removeListeners": [Function],
  "requestAuthorization": undefined,
  "startRecording": [Function],
  "stopRecording": [Function],
  "unbindNativeEventListeners": [Function],
}
`;

exports[`Recorder removeListeners recordingProgressSubscription 1`] = `
Object {
  "bindNativeEventListeners": [Function],
  "checkAuthorizationStatus": undefined,
  "destroy": [Function],
  "init": [Function],
  "onRecordingFinished": null,
  "onRecordingProgress": null,
  "pauseRecording": [Function],
  "removeListeners": [Function],
  "requestAuthorization": undefined,
  "startRecording": [Function],
  "stopRecording": [Function],
  "unbindNativeEventListeners": [Function],
}
`;
